tun/tun_illumos.go

```diff
@@ -0,0 +1,519 @@
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+func getmsg(fd int, ctlptr uintptr, dataptr uintptr, flagsp uintptr) (
+    r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_getmsg)), 4,
+	    uintptr(fd), ctlptr, dataptr, flagsp, 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+
+/*
+ * WIREGUARD/TUN INTERFACE
+ */
+
+type NativeTun struct {
+	tunFile *os.File	/* TUN device file */
+	ip_fd int		/* IP device fd */
+	name string		/* Interface name */
+	mtu int
+
+	events chan Event
+	errors chan error
+}
+
+// type Device interface {
+// 	File() *os.File                 // returns the file descriptor of the device
+// 	Read([]byte, int) (int, error)  // read a packet from the device (without any additional headers)
+// 	Write([]byte, int) (int, error) // writes a packet to the device (without any additional headers)
+// 	MTU() (int, error)              // returns the MTU of the device
+// 	Name() (string, error)          // fetches and returns the current name
+// 	Events() chan Event          // returns a constant channel of events related to the device
+// 	Close() error                   // stops the device and closes the event channel
+// }
+
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+func (tun *NativeTun) Flush() error {
+        // TODO: can flushing be implemented by buffering and using sendmmsg?
+        return nil
+}
+
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+
+	default:
+		return tun.read_tun(buf[offset:])
+	}
+}
+
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	return tun.write_tun(buf[offset:])
+}
+
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+func (tun *NativeTun) MTU() (int, error) {
+	return tun.mtu, nil
+}
+
+func (tun *NativeTun) Events() chan Event {
+	return tun.events
+}
+
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
+		ip_muxid, err := get_ip_muxid(tun.ip_fd, tun.name)
+		if err != nil {
+			return err
+		}
+
+		err = punlink(tun.ip_fd, ip_muxid)
+		if err != nil {
+			return err
+		}
+
+		unix.Close(tun.ip_fd)
+		tun.ip_fd = -1
+	}
+
+	if tun.tunFile != nil {
+		tun.tunFile.Close()
+		tun.tunFile = nil
+	}
+
+	if tun.events != nil {
+		close(tun.events)
+	}
+
+	return nil
+}
+
+func CreateTUNFromFile(file *os.File, mtu int) (Device, error) {
+	/*
+	 * XXX It's not currently clear to me how to take a TUN file descriptor
+	 * and determine the attached PPA.
+	 *
+	 * XXX Based on the current architecture of the daemon, we need to be
+	 * able to reconstitute our NativeTun object from just a file
+	 * descriptor number as the fd number is passed (with no other
+	 * information) through the environment to the child.  For now, use the
+	 * "-f" flag.
+	 */
+	return nil, fmt.Errorf("CreateTUNFromFile() not currently supported")
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
```

Comment:
I presume this is because there's no way of requesting a specific interface name, so you hardcoded this backstop in? If we can track down if_tun.h (or the corresponding kernel side implementation), I expect it'll have an ioctl for setting names as well, at which point we can remove this.

Reply:
I'd be surprised if at least tunX wasn't allowed. Can you dig around and become certain one way or another?

Reply:
Any update here?

Reply:
tunX is allowed and now implemented.